File: scala/day02.scala
//> using toolkit default

def base(pred: String => Boolean): Long =
  os.read(os.pwd / "input" / "day02" / "in.txt")
    .strip
    .split(",")
    .map(_.trim)
    .collect { case s"$start-$end" => start.toLong to end.toLong }
    .flatMap(range =>
      range.view
        .map(_.toString)
        .filter(pred)
        .map(_.toLong)
    )
    .sum

def partOne(): Long =
  base(s =>
    val mid = s.length / 2
    s.length % 2 == 0 && s.take(mid) == s.drop(mid)
  )

def partTwo(): Long =
  base(s =>
    (1 to s.length / 2)
      .filter(gs => s.length % gs == 0)
      .exists(gs =>
        val seq = s.take(gs)
        s.grouped(gs).forall(_ == seq)
      )
  )

@main def run(): Unit =
  println(partOne())
  println(partTwo())

File: python/day02.py
def p1():
    sum = 0
    with open("input/day02/in.txt") as f:
        line = f.read().strip()
        rs = line.split(",")
        for r in rs:
            r = r.split("-")
            f = int(r[0])
            l = int(r[1])
            for id in range(f, l+1):
                sid = str(id)
                if len(sid) % 2 == 0 and sid[:len(sid)//2] == sid[len(sid)//2:]:
                    sum += id
    print(sum)

def p2():
    sum = 0
    with open("input/day02/in.txt") as f:
        line = f.read().strip()
        rs = line.split(",")
        for r in rs:
            r = r.split("-")
            f = int(r[0])
            l = int(r[1])
            for id in range(f, l+1):
                sid = str(id)
                for s in range(1, len(sid)//2 + 1):
                    windows = []
                    for i in range(0, len(sid), s):
                        windows.append(sid[i:i+s])

                    if all(window == windows[0] for window in windows):
                        sum += id
                        break
    print(sum)

if __name__ == '__main__':
    p1()
    p2()

